packages/postgres-database/tests/projects/test_projects_version_control.py

```diff
@@ -0,0 +1,285 @@
+        checksum=checksum,
+        content={"workbench": project_wc.workbench, "ui": project_wc.ui},
+    )
+    assert row_id == checksum
+    return checksum
+
+
+async def test_basic_workflow(project: RowProxy, conn: SAConnection):
+
+    # git init
+    async with conn.begin():
+        # create repo
+        repo_orm = ReposOrm(conn)
+        repo_id = await repo_orm.insert(project_uuid=project.uuid)
+        assert repo_id is not None
+
+        repo_orm.pin(rowid=repo_id)
+        repo = await repo_orm.fetch()
+        assert repo
+        assert repo.project_uuid == project.uuid
+        assert repo.project_checksum is None
+        assert repo.created == repo.modified
+
+        # create main branch
+        branches_orm = BranchesOrm(conn)
+        branch_id = await branches_orm.insert(repo_id=repo.id)
+        assert branch_id is not None
+
+        branches_orm.pin(rowid=branch_id)
+        main_branch: Optional[RowProxy] = await branches_orm.fetch()
+        assert main_branch
+        assert main_branch.name == "main", "Expected 'main' as default branch"
+        assert main_branch.head_commit_id is None, "still not assigned"
+        assert main_branch.created == main_branch.modified
+
+        # assign head branch
+        heads_orm = HeadsOrm(conn)
+        await heads_orm.insert(repo_id=repo.id, head_branch_id=branch_id)
+
+        heads_orm.pin(rowid=repo.id)
+        head = await heads_orm.fetch()
+        assert head
+
+    #
+    # create first commit -- TODO: separate tests
+
+    # fetch a *full copy* of the project (WC)
+    repo = await repo_orm.fetch("id project_uuid project_checksum")
+    assert repo
+
+    project_orm = ProjectsOrm(conn).pin(uuid=repo.project_uuid)
+    project_wc = await project_orm.fetch()
+    assert project_wc
+    assert project == project_wc
+
+    # eval checksum
+    checksum = eval_checksum(project_wc.workbench)
+    assert repo.project_checksum != checksum
+
+    # take snapshot <=> git add & commit
+    async with conn.begin():
+
+        snapshot_checksum = await add_snapshot(project_wc, checksum, repo, conn)
+
+        # get HEAD = repo.branch_id -> .head_commit_id
+        assert head.repo_id == repo.id
+        branches_orm.pin(head.head_branch_id)
+        branch = await branches_orm.fetch("head_commit_id name")
+        assert branch
+        assert branch.name == "main"
+        assert branch.head_commit_id is None, "First commit"
+
+        # create commit
+        commits_orm = CommitsOrm(conn)
+        commit_id = await commits_orm.insert(
+            repo_id=repo.id,
+            parent_commit_id=branch.head_commit_id,
+            snapshot_checksum=snapshot_checksum,
+            message="first commit",
+        )
+        assert commit_id
+        assert isinstance(commit_id, int)
+
+        # update branch head
+        await branches_orm.update(head_commit_id=commit_id)
+
+        # update checksum cache
+        await repo_orm.update(project_checksum=snapshot_checksum)
```

Comment:
I hope you will have some wrappers around these calls... ;) that are a lot of insert, update, to do in the right order.

Reply:
so, my idea is to extend this to [subqueries](https://docs.sqlalchemy.org/en/14/orm/queryguide.html#joining-to-subqueries) such that I can compose several and just do one execute call

Reply:
For the future, I'd also prefer having a nice interface hiding all this.

Reply:
yes, of course, these will be used inside the a ``BaseRepository`` subclass in the ``webserver``